pkg/controller/populators/upload-populator.go

```diff
@@ -0,0 +1,231 @@
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := addCommonPopulatorsWatches(mgr, uploadPopulator); err != nil {
+		return nil, err
+	}
+	if err := addUploadPopulatorWatches(mgr, uploadPopulator, log); err != nil {
+		return nil, err
+	}
+	return uploadPopulator, nil
+}
+
+func addUploadPopulatorWatches(mgr manager.Manager, c controller.Controller, log logr.Logger) error {
+	const dataSourceRefField = "spec.dataSourceRef"
+
+	indexKeyFunc := func(namespace, name string) string {
+		return namespace + "/" + name
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &corev1.PersistentVolumeClaim{}, dataSourceRefField, func(obj client.Object) []string {
+		if dsr := obj.(*corev1.PersistentVolumeClaim).Spec.DataSourceRef; dsr != nil {
+			return []string{indexKeyFunc(obj.GetNamespace(), dsr.Name)}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	mapUploadSourceToPVC := func(obj client.Object) (reqs []reconcile.Request) {
+		var pvcs corev1.PersistentVolumeClaimList
+		matchingFields := client.MatchingFields{dataSourceRefField: indexKeyFunc(obj.GetNamespace(), obj.GetName())}
+		if err := mgr.GetClient().List(context.TODO(), &pvcs, matchingFields); err != nil {
+			log.Error(err, "Unable to list PVCs", "matchingFields", matchingFields)
+			return reqs
+		}
+		for _, pvc := range pvcs.Items {
+			reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}})
+		}
+		return
+	}
+
+	if err := c.Watch(&source.Kind{Type: &cdiv1.UploadSource{}},
+		handler.EnqueueRequestsFromMapFunc(mapUploadSourceToPVC),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile the reconcile loop for the PVC with DataSourceRef of UploadSource kind
+func (r *UploadPopulatorReconciler) Reconcile(_ context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("PVC", req.NamespacedName)
+	log.V(1).Info("reconciling Upload Source PVCs")
+
+	// Get the PVC.
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), req.NamespacedName, pvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	dataSourceRef := pvc.Spec.DataSourceRef
+	if dataSourceRef != nil {
+		if !IsPVCDataSourceRefKind(pvc, cdiv1.UploadSourceRef) || dataSourceRef.Name == "" {
+			// Ignore PVCs that aren't for this populator to handle
+			return reconcile.Result{}, nil
+		}
+		if exists, err := r.verifyUploadSourceExists(pvc.Namespace, dataSourceRef.Name); err != nil || !exists {
+			return reconcile.Result{}, err
+		}
+		return r.reconcileTargetPVC(pvc)
+	}
+
+	if isPVCPrime(pvc) {
+		_, ok := pvc.Annotations[cc.AnnUploadRequest]
+		if !ok {
+			// Ignore PVCs that aren't for this populator to handle
+			return reconcile.Result{}, nil
+		}
+		return r.reconcilePVCPrime(pvc)
```

Comment:
it is cleaner to just reconcile target pvc and have the watch will trigger target pvc reconcile when pvc prime changes

Reply:
Populator should not care about the upload pod.  THe upload PVC (pvc prime) has all relevant status